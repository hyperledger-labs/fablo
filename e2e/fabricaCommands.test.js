const { execSync } = require('child_process');
const currentFabricaVersion = require('../package.json').version;

const commandOutput = (status, output) => ({
  status,
  output,
  outputJson: () => JSON.parse(output),
});

const executeCommand = (c, noConsole = false) => {
  // eslint-disable-next-line no-console
  const log = !noConsole ? (out) => console.log(out) : () => {};
  try {
    log(c);
    const output = execSync(c, { encoding: 'utf-8' });
    log(output);
    return commandOutput(0, output);
  } catch (e) {
    const output = (e.output || []).join('');
    // eslint-disable-next-line no-console
    console.error(output);
    return commandOutput(e.status, output);
  }
};

const success = () => expect.objectContaining({ status: 0 });
const failure = () => expect.objectContaining({ status: 1 });

const workdir = './e2e/__tmp__/commands-tests';
const relativeRoot = '../../..';

const fabricaExec = (command) => executeCommand(`cd ${workdir} && ${relativeRoot}/fabrica.sh ${command}`);

const getFiles = () => executeCommand(`find '${workdir}' -type f`)
  .output
  .split('\n')
  .filter((s) => !!s.length)
  .sort();

const getFileContent = (file) => executeCommand(`cat ${file}`).output;

const cleanupWorkdir = () => {
  execSync(`rm -rf ${workdir} ; mkdir -p ${workdir}`);
};

describe('init', () => {
  beforeEach(cleanupWorkdir);

  it('should init simple fabrica config', () => {
    // When
    const commandResult = fabricaExec('init');

    // Then
    expect(commandResult).toEqual(success());
    expect(commandResult.output).toContain('Sample config file created! :)');
    expect(commandResult.output).toContain('Chaincode directory is \'./chaincodes/chaincode-kv-node\'.');
    expect(commandResult.output).toContain('If it\'s empty your network won\'t run entirely.');
    expect(getFiles()).toEqual(['./e2e/__tmp__/commands-tests/fabrica-config.json']);
    expect(getFileContent(`${workdir}/fabrica-config.json`)).toMatchSnapshot();
  });
});

describe('use', () => {
  beforeEach(cleanupWorkdir);

  it('should display versions', () => {
    // When
    const commandResult = fabricaExec('use');

    // Then
    expect(commandResult).toEqual(success());
    expect(commandResult.output).toContain('0.0.1');
    expect(commandResult.output).toContain(`${currentFabricaVersion} <== current`);
    expect(getFiles()).toEqual([]);
  });
});

describe('validate', () => {
  beforeEach(cleanupWorkdir);

  it('should validate default config', () => {
    // Given
    fabricaExec('init');

    // When
    const commandResult = fabricaExec('validate');

    // Then
    expect(commandResult).toEqual(success());
    expect(commandResult.output).toContain('Validation errors count: 0');
    expect(commandResult.output).toContain('Validation warnings count: 0');
    expect(getFiles()).toEqual(['./e2e/__tmp__/commands-tests/fabrica-config.json']);
  });

  it('should validate custom config', () => {
    // Given
    const fabricaConfig = `${relativeRoot}/samples/fabricaConfig-2orgs-2channels-2chaincodes-tls-raft.json`;

    // When
    const commandResult = fabricaExec(`validate ${fabricaConfig}`);

    // Then
    expect(commandResult).toEqual(success());
    expect(commandResult.output).toContain('Validation errors count: 0');
    expect(commandResult.output).toContain('Validation warnings count: 0');
    expect(getFiles()).toEqual([]);
  });

  it('should fail to validate if config file is missing', () => {
    const commandResult = fabricaExec('validate');

    // Then
    expect(commandResult).toEqual(failure());
    expect(commandResult.output).toContain('commands-tests/fabrica-config.json does not exist\n');
    expect(getFiles()).toEqual([]);
  });
});

describe('version', () => {
  it('should print version information', () => {
    // When
    const commandResult = fabricaExec('version');

    // Then
    expect(commandResult).toEqual(success());
    expect(commandResult.outputJson()).toEqual(expect.objectContaining({
      version: currentFabricaVersion,
      build: expect.stringMatching(/.*/),
    }));
  });

  it('should print verbose version information', () => {
    // When
    const commandResult1 = fabricaExec('version -v');
    const commandResult2 = fabricaExec('version --verbose');

    // Then
    expect(commandResult1).toEqual(success());
    expect(commandResult1.outputJson()).toEqual(expect.objectContaining({
      version: currentFabricaVersion,
      build: expect.stringMatching(/.*/),
      supported: expect.objectContaining({
        fabricaVersions: expect.stringMatching(/.*/),
        hyperledgerFabricVersions: expect.anything(),
      }),
    }));

    expect(commandResult1.status).toEqual(commandResult2.status);
    expect(commandResult1.output).toEqual(commandResult2.output);
  });
});

// TODO describe(printConfig)
