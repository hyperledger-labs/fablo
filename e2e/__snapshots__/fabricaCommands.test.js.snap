// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extend config should extend custom config 1`] = `
"Fabrica is powered by :
 _____        __ _                         ___  ____ _ _ 
/  ___|      / _| |                        |  \\\\/  (_) | |
\\\\ \`--.  ___ | |_| |___      ____ _ _ __ ___| .  . |_| | |
 \`--. \\\\/ _ \\\\|  _| __\\\\ \\\\ /\\\\ / / _\` | '__/ _ \\\\ |\\\\/| | | | |
/\\\\__/ / (_) | | | |_ \\\\ V  V / (_| | | |  __/ |  | | | | |
\\\\____/ \\\\___/|_|  \\\\__| \\\\_/\\\\_/ \\\\__,_|_|  \\\\___\\\\_|  |_/_|_|_|
=========================================================== v: 0.1.0-unstable
Validation errors count: 0
Validation warnings count: 0
===========================================================
"
`;

exports[`extend config should extend default config 1`] = `
"Fabrica is powered by :
 _____        __ _                         ___  ____ _ _ 
/  ___|      / _| |                        |  \\\\/  (_) | |
\\\\ \`--.  ___ | |_| |___      ____ _ _ __ ___| .  . |_| | |
 \`--. \\\\/ _ \\\\|  _| __\\\\ \\\\ /\\\\ / / _\` | '__/ _ \\\\ |\\\\/| | | | |
/\\\\__/ / (_) | | | |_ \\\\ V  V / (_| | | |  __/ |  | | | | |
\\\\____/ \\\\___/|_|  \\\\__| \\\\_/\\\\_/ \\\\__,_|_|  \\\\___\\\\_|  |_/_|_|_|
=========================================================== v: 0.1.0-unstable
Validation errors count: 0
Validation warnings count: 0
===========================================================
{
  \\"networkSettings\\": {
    \\"fabricVersion\\": \\"1.4.6\\",
    \\"tls\\": false,
    \\"fabricCaVersion\\": \\"1.4.6\\",
    \\"paths\\": {
      \\"fabricaConfig\\": \\"<absolute path>\\",
      \\"chaincodesBaseDir\\": \\"<absolute path>\\"
    },
    \\"isHlf20\\": false
  },
  \\"capabilities\\": {
    \\"channel\\": \\"V1_4_3\\",
    \\"orderer\\": \\"V1_4_2\\",
    \\"application\\": \\"V1_4_2\\"
  },
  \\"rootOrg\\": {
    \\"name\\": \\"Orderer\\",
    \\"mspName\\": \\"OrdererMSP\\",
    \\"domain\\": \\"root.com\\",
    \\"organization\\": {
      \\"key\\": \\"root\\",
      \\"name\\": \\"Orderer\\",
      \\"mspName\\": \\"OrdererMSP\\",
      \\"domain\\": \\"root.com\\"
    },
    \\"ca\\": {
      \\"prefix\\": \\"ca\\",
      \\"address\\": \\"ca.root.com\\",
      \\"port\\": 7054,
      \\"exposePort\\": 7030,
      \\"fullAddress\\": \\"ca.root.com:7054\\",
      \\"caAdminNameVar\\": \\"ORDERER_CA_ADMIN_NAME\\",
      \\"caAdminPassVar\\": \\"ORDERER_CA_ADMIN_PASSWORD\\"
    },
    \\"orderers\\": [
      {
        \\"name\\": \\"orderer0\\",
        \\"domain\\": \\"root.com\\",
        \\"address\\": \\"orderer0.root.com\\",
        \\"consensus\\": \\"solo\\",
        \\"port\\": 7050,
        \\"fullAddress\\": \\"orderer0.root.com:7050\\"
      }
    ],
    \\"ordererHead\\": {
      \\"name\\": \\"orderer0\\",
      \\"domain\\": \\"root.com\\",
      \\"address\\": \\"orderer0.root.com\\",
      \\"consensus\\": \\"solo\\",
      \\"port\\": 7050,
      \\"fullAddress\\": \\"orderer0.root.com:7050\\"
    }
  },
  \\"orgs\\": [
    {
      \\"key\\": \\"org1\\",
      \\"name\\": \\"Org1\\",
      \\"mspName\\": \\"Org1MSP\\",
      \\"domain\\": \\"org1.com\\",
      \\"cryptoConfigFileName\\": \\"crypto-config-org1\\",
      \\"peersCount\\": 2,
      \\"peers\\": [
        {
          \\"name\\": \\"peer0\\",
          \\"address\\": \\"peer0.org1.com\\",
          \\"db\\": \\"LevelDb\\",
          \\"isAnchorPeer\\": true,
          \\"port\\": 7060,
          \\"fullAddress\\": \\"peer0.org1.com:7060\\",
          \\"couchDbExposePort\\": 5080
        },
        {
          \\"name\\": \\"peer1\\",
          \\"address\\": \\"peer1.org1.com\\",
          \\"db\\": \\"LevelDb\\",
          \\"isAnchorPeer\\": false,
          \\"port\\": 7061,
          \\"fullAddress\\": \\"peer1.org1.com:7061\\",
          \\"couchDbExposePort\\": 5081
        }
      ],
      \\"anchorPeers\\": [
        {
          \\"name\\": \\"peer0\\",
          \\"address\\": \\"peer0.org1.com\\",
          \\"db\\": \\"LevelDb\\",
          \\"isAnchorPeer\\": true,
          \\"port\\": 7060,
          \\"fullAddress\\": \\"peer0.org1.com:7060\\",
          \\"couchDbExposePort\\": 5080
        }
      ],
      \\"bootstrapPeers\\": \\"peer0.org1.com:7060\\",
      \\"ca\\": {
        \\"prefix\\": \\"ca\\",
        \\"address\\": \\"ca.org1.com\\",
        \\"port\\": 7054,
        \\"exposePort\\": 7031,
        \\"fullAddress\\": \\"ca.org1.com:7054\\",
        \\"caAdminNameVar\\": \\"ORG1_CA_ADMIN_NAME\\",
        \\"caAdminPassVar\\": \\"ORG1_CA_ADMIN_PASSWORD\\"
      },
      \\"headPeer\\": {
        \\"name\\": \\"peer0\\",
        \\"address\\": \\"peer0.org1.com\\",
        \\"db\\": \\"LevelDb\\",
        \\"isAnchorPeer\\": true,
        \\"port\\": 7060,
        \\"fullAddress\\": \\"peer0.org1.com:7060\\",
        \\"couchDbExposePort\\": 5080
      }
    }
  ],
  \\"channels\\": [
    {
      \\"key\\": \\"channel1\\",
      \\"name\\": \\"my-channel1\\",
      \\"orgs\\": [
        {
          \\"name\\": \\"Org1\\",
          \\"mspName\\": \\"Org1MSP\\",
          \\"domain\\": \\"org1.com\\",
          \\"peers\\": [
            {
              \\"name\\": \\"peer0\\",
              \\"address\\": \\"peer0.org1.com\\",
              \\"db\\": \\"LevelDb\\",
              \\"isAnchorPeer\\": true,
              \\"port\\": 7060,
              \\"fullAddress\\": \\"peer0.org1.com:7060\\",
              \\"couchDbExposePort\\": 5080
            },
            {
              \\"name\\": \\"peer1\\",
              \\"address\\": \\"peer1.org1.com\\",
              \\"db\\": \\"LevelDb\\",
              \\"isAnchorPeer\\": false,
              \\"port\\": 7061,
              \\"fullAddress\\": \\"peer1.org1.com:7061\\",
              \\"couchDbExposePort\\": 5081
            }
          ],
          \\"headPeer\\": {
            \\"name\\": \\"peer0\\",
            \\"address\\": \\"peer0.org1.com\\",
            \\"db\\": \\"LevelDb\\",
            \\"isAnchorPeer\\": true,
            \\"port\\": 7060,
            \\"fullAddress\\": \\"peer0.org1.com:7060\\",
            \\"couchDbExposePort\\": 5080
          }
        }
      ],
      \\"profile\\": {
        \\"name\\": \\"MyChannel1\\"
      },
      \\"instantiatingOrg\\": {
        \\"name\\": \\"Org1\\",
        \\"mspName\\": \\"Org1MSP\\",
        \\"domain\\": \\"org1.com\\",
        \\"peers\\": [
          {
            \\"name\\": \\"peer0\\",
            \\"address\\": \\"peer0.org1.com\\",
            \\"db\\": \\"LevelDb\\",
            \\"isAnchorPeer\\": true,
            \\"port\\": 7060,
            \\"fullAddress\\": \\"peer0.org1.com:7060\\",
            \\"couchDbExposePort\\": 5080
          },
          {
            \\"name\\": \\"peer1\\",
            \\"address\\": \\"peer1.org1.com\\",
            \\"db\\": \\"LevelDb\\",
            \\"isAnchorPeer\\": false,
            \\"port\\": 7061,
            \\"fullAddress\\": \\"peer1.org1.com:7061\\",
            \\"couchDbExposePort\\": 5081
          }
        ],
        \\"headPeer\\": {
          \\"name\\": \\"peer0\\",
          \\"address\\": \\"peer0.org1.com\\",
          \\"db\\": \\"LevelDb\\",
          \\"isAnchorPeer\\": true,
          \\"port\\": 7060,
          \\"fullAddress\\": \\"peer0.org1.com:7060\\",
          \\"couchDbExposePort\\": 5080
        }
      }
    }
  ],
  \\"chaincodes\\": [
    {
      \\"directory\\": \\"./chaincodes/chaincode-kv-node\\",
      \\"name\\": \\"chaincode1\\",
      \\"version\\": \\"0.0.1\\",
      \\"lang\\": \\"node\\",
      \\"channel\\": {
        \\"key\\": \\"channel1\\",
        \\"name\\": \\"my-channel1\\",
        \\"orgs\\": [
          {
            \\"name\\": \\"Org1\\",
            \\"mspName\\": \\"Org1MSP\\",
            \\"domain\\": \\"org1.com\\",
            \\"peers\\": [
              {
                \\"name\\": \\"peer0\\",
                \\"address\\": \\"peer0.org1.com\\",
                \\"db\\": \\"LevelDb\\",
                \\"isAnchorPeer\\": true,
                \\"port\\": 7060,
                \\"fullAddress\\": \\"peer0.org1.com:7060\\",
                \\"couchDbExposePort\\": 5080
              },
              {
                \\"name\\": \\"peer1\\",
                \\"address\\": \\"peer1.org1.com\\",
                \\"db\\": \\"LevelDb\\",
                \\"isAnchorPeer\\": false,
                \\"port\\": 7061,
                \\"fullAddress\\": \\"peer1.org1.com:7061\\",
                \\"couchDbExposePort\\": 5081
              }
            ],
            \\"headPeer\\": {
              \\"name\\": \\"peer0\\",
              \\"address\\": \\"peer0.org1.com\\",
              \\"db\\": \\"LevelDb\\",
              \\"isAnchorPeer\\": true,
              \\"port\\": 7060,
              \\"fullAddress\\": \\"peer0.org1.com:7060\\",
              \\"couchDbExposePort\\": 5080
            }
          }
        ],
        \\"profile\\": {
          \\"name\\": \\"MyChannel1\\"
        },
        \\"instantiatingOrg\\": {
          \\"name\\": \\"Org1\\",
          \\"mspName\\": \\"Org1MSP\\",
          \\"domain\\": \\"org1.com\\",
          \\"peers\\": [
            {
              \\"name\\": \\"peer0\\",
              \\"address\\": \\"peer0.org1.com\\",
              \\"db\\": \\"LevelDb\\",
              \\"isAnchorPeer\\": true,
              \\"port\\": 7060,
              \\"fullAddress\\": \\"peer0.org1.com:7060\\",
              \\"couchDbExposePort\\": 5080
            },
            {
              \\"name\\": \\"peer1\\",
              \\"address\\": \\"peer1.org1.com\\",
              \\"db\\": \\"LevelDb\\",
              \\"isAnchorPeer\\": false,
              \\"port\\": 7061,
              \\"fullAddress\\": \\"peer1.org1.com:7061\\",
              \\"couchDbExposePort\\": 5081
            }
          ],
          \\"headPeer\\": {
            \\"name\\": \\"peer0\\",
            \\"address\\": \\"peer0.org1.com\\",
            \\"db\\": \\"LevelDb\\",
            \\"isAnchorPeer\\": true,
            \\"port\\": 7060,
            \\"fullAddress\\": \\"peer0.org1.com:7060\\",
            \\"couchDbExposePort\\": 5080
          }
        }
      },
      \\"init\\": \\"{\\\\\\"Args\\\\\\":[]}\\",
      \\"endorsement\\": \\"AND ('Org1MSP.member')\\",
      \\"instantiatingOrg\\": {
        \\"name\\": \\"Org1\\",
        \\"mspName\\": \\"Org1MSP\\",
        \\"domain\\": \\"org1.com\\",
        \\"peers\\": [
          {
            \\"name\\": \\"peer0\\",
            \\"address\\": \\"peer0.org1.com\\",
            \\"db\\": \\"LevelDb\\",
            \\"isAnchorPeer\\": true,
            \\"port\\": 7060,
            \\"fullAddress\\": \\"peer0.org1.com:7060\\",
            \\"couchDbExposePort\\": 5080
          },
          {
            \\"name\\": \\"peer1\\",
            \\"address\\": \\"peer1.org1.com\\",
            \\"db\\": \\"LevelDb\\",
            \\"isAnchorPeer\\": false,
            \\"port\\": 7061,
            \\"fullAddress\\": \\"peer1.org1.com:7061\\",
            \\"couchDbExposePort\\": 5081
          }
        ],
        \\"headPeer\\": {
          \\"name\\": \\"peer0\\",
          \\"address\\": \\"peer0.org1.com\\",
          \\"db\\": \\"LevelDb\\",
          \\"isAnchorPeer\\": true,
          \\"port\\": 7060,
          \\"fullAddress\\": \\"peer0.org1.com:7060\\",
          \\"couchDbExposePort\\": 5080
        }
      },
      \\"privateData\\": []
    }
  ]
}
"
`;

exports[`init should init simple fabrica config 1`] = `
"{
  \\"$schema\\": \\"https://github.com/softwaremill/fabrica/releases/download/0.1.0-unstable/schema.json\\",
  \\"networkSettings\\": {
    \\"fabricVersion\\": \\"1.4.6\\",
    \\"tls\\": false
  },
  \\"rootOrg\\": {
    \\"organization\\": {
      \\"key\\": \\"root\\",
      \\"name\\": \\"Orderer\\",
      \\"mspName\\": \\"OrdererMSP\\",
      \\"domain\\": \\"root.com\\"
    },
    \\"ca\\": {
      \\"prefix\\": \\"ca\\"
    },
    \\"orderer\\": {
      \\"prefix\\": \\"orderer\\",
      \\"type\\": \\"solo\\",
      \\"instances\\": 1
    }
  },
  \\"orgs\\": [
    {
      \\"organization\\": {
        \\"key\\": \\"org1\\",
        \\"name\\": \\"Org1\\",
        \\"mspName\\": \\"Org1MSP\\",
        \\"domain\\": \\"org1.com\\"
      },
      \\"ca\\": {
        \\"prefix\\": \\"ca\\"
      },
      \\"peer\\": {
        \\"prefix\\": \\"peer\\",
        \\"instances\\": 2,
        \\"db\\": \\"LevelDb\\"
      }
    }
  ],
  \\"channels\\": [
    {
      \\"key\\": \\"channel1\\",
      \\"name\\": \\"my-channel1\\",
      \\"orgs\\": [
        {
          \\"key\\": \\"org1\\",
          \\"peers\\": [
            \\"peer0\\",
            \\"peer1\\"
          ]
        }
      ]
    }
  ],
  \\"chaincodes\\": [
    {
      \\"name\\": \\"chaincode1\\",
      \\"version\\": \\"0.0.1\\",
      \\"lang\\": \\"node\\",
      \\"channel\\": \\"channel1\\",
      \\"init\\": \\"{\\\\\\"Args\\\\\\":[]}\\",
      \\"endorsement\\": \\"AND ('Org1MSP.member')\\",
      \\"directory\\": \\"./chaincodes/chaincode-kv-node\\"
    }
  ]
}
"
`;
